### PR TITLE
Llama lora pure torch fine tuning

### PR DESCRIPTION
**Pull request by pmarkovicTT**
### Ticket
closes: https://github.com/tenstorrent/tt-blacksmith/issues/132 https://github.com/tenstorrent/tt-blacksmith/issues/124 https://github.com/tenstorrent/tt-blacksmith/issues/125 https://github.com/tenstorrent/tt-blacksmith/issues/129 https://github.com/tenstorrent/tt-blacksmith/issues/123

### Problem description
We need a training script for Llama lora fine tuning with pure PyTorch so we can have fair comparison with GPU training.

### What's changed
Standard PyTorch training loop

### Checklist
- [ ] Add script for pure torch llama lora training

### CODE REVIEW

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                    torch_optimizer.step()
+                    torch_optimizer.zero_grad()
+
+                global_step += 1
+                if global_step % log_every_n_steps == 0:
+                    avg_loss = running_loss / log_every_n_steps
+                    run.log({"train/loss": avg_loss, "step": global_step})
+                    running_loss = 0.0
+
+            if config.save_strategy == "epoch":
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py:78]: what if save_strategy isn't 'epoch' - what are other options? We can develop this a bit; I'm guessing the other options is sth similar to log_every_n_steps, like 'save_every_n_steps'.
We don't want to save a checkpoint probably only if we explicitly set it through a specific save_strategy flag.
Let me know if I'm missing something.

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    config_file_path = os.path.join(os.path.dirname(__file__), "test_llama_fine_tuning_pure_torch.yaml")
+    config = generate_config(TrainingConfig, config_file_path)
+
+    os.makedirs(os.path.join(config.output_dir, "checkpoints"), exist_ok=True)
+
+    model = get_model(config)
+
+    dataset = SSTDataset(config)
+    train_set, eval_set = dataset.load_tokenized_data()
+    train_data_loader = DataLoader(train_set, batch_size=config.batch_size, shuffle=False, drop_last=True)
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py:115]: I think we should have shuffle=True in train_data_loader.

File: blacksmith/tools/torch_helpers.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        total_params += param_size
+
+        if param.requires_grad:
+            total_grads += param_size
+
+    # Calculate buffer size (non-parameters that require memory)
+    total_buffers = sum(buf.numel() for buf in model.buffers())
+
+    # Size in bytes = (Number of elements) * (Size of each element in bytes)
+    # We assume parameters and gradients are stored in the same type as input dtype
```
Comment (vladimirjovanovicTT) [inline blacksmith/tools/torch_helpers.py:35]: Optimally we would loop through parameters and gather their dtype but I think this is ok for now.

File: blacksmith/tools/torch_helpers.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    print(f"Input dtype: {input_dtype}")
+    print(f"Model size: {total_memory_gb:.2f} GB")
+    print(f"Parameters: {total_params} | Gradients: {total_grads} | Buffers: {total_buffers}")
+
+    return total_memory_gb
+
+
+def log_mem(stage):
+    allocated = torch.cuda.memory_allocated() / 1e9
```
Comment (vladimirjovanovicTT) [inline blacksmith/tools/torch_helpers.py:50]: nit: we are using 1e9 here and 1024^3 above.

**Review by vladimirjovanovicTT — APPROVED** [review]
Looks good, several comments.

File: blacksmith/datasets/torch/llama/sst_dataset.py [hunk 5]
```diff
@@ -4,8 +4,9 @@
 from typing import Tuple, Dict, Any
 from datasets import load_dataset
 from transformers import AutoTokenizer
+import torch
```
[the remark this answers is not in the document]
Comment (pmarkovicTT) [reply blacksmith/datasets/torch/llama/sst_dataset.py:7]: no that I know, we didn't update hooks recently
